### DISCUSSION TITLE
[RFC] Add construct_inputs class method to acquisition functions

### DISCUSSION THREAD

**Discussion by Balandat** [meta-pytorch/botorch#784] [Ideas]
**Context:**
#783 introduces a new `construct_inputs` class method on `AcquisitionFunction`s that can be used to construct the arguments to the respective acquisition function constructor in a programmatic way based on a general set of inputs. This is analogous to the `construct_inputs` methods on the `Model`s that we already have in place. It would be an optional method and when called on acquisition functions that do not implement custom handling would just raise a `NotImplementedError`. 

**Why do this:**
The main use case for these is to simplify the instantiation of acquisition functions based on a general set of inputs with a consistent call signature in higher-level libraries using botorch as a modeling / optimization engine under the hood. Specifically, this would significantly simplify putting together custom combinations of Surrogate models, Acquisition functions, and optimizers in the modular botorch model setup in Ax: https://github.com/facebook/Ax/tree/master/ax/models/torch/botorch_modular. That avoids having to write a big old registry in Ax that is not only ugly but will also be very hard to keep in sync with the BoTorch changes.

**Why not do this:**
This adds a bunch of additional complexity and code that's not actually used directly within BoTorch. This means that even though this functionality is optional, the more complicated implementations of our stock acquisition functions could scare away users from using BoTorch in the first place. It also means additional maintenance overhead. 

I am curious to hear what people think about this proposal.

**Comment by Balandat** [comment]
@sdaulton @danielrjiang @dme65 @qingfeng10 @bletham @eytan @bkarrer @liangshi7 @mshvartsman @saitcakmak @wjmaddox @henrymoss, any thoughts on this? Specifically about whether you think this additional complexity / overhead would make it harder (or at lest perceived to be harder) to work with BoTorch.

**Reply by sdaulton** [reply]
I do think construct_inputs is a bit confusing. Currently, many objects are pulled out of kwargs, which requires reading through the modular botorch model in Ax to figure out what is passed in to construct_inputs. This is kind of hard to read on the Ax side too because many things are also passed around via **kwargs. One other thought on construct_inputs: the functionality is not too different from `get_acquisition_function` (https://github.com/pytorch/botorch/blob/master/botorch/acquisition/utils.py#L35). Is there benefit to putting this construct_inputs method on the class as opposed to in a utility? The benefit of a utility is that is does not overcomplicate the core acquisition function classes, with something that is not used in BoTorch and really requires reading Ax code to digest.

**Reply by Balandat** [reply]
> Is there benefit to putting this construct_inputs method on the class as opposed to in a utility?

The primary benefits are 
1. class constructor and utility for constructing constructor args are co-located, so it's easier to keep track of things
2. we avoid a pretty ugly rats tail of conditional statements, with logic that is likely to break and likely to be overlooked when making changes to the acquisition functions.

**Reply by Balandat** [reply]
Re 2.: We can probably make this some kind of dict-based dispatch to avoid the rats tail and keep things reasonably organized. I'll prototype sth for comparison.

**Reply by lena-kashtelyan** [reply]
I think a comparison would be very helpful! I agree that a utility could be better and would complicate classes less, but could also be a bit more brittle over time, similar to `botorch_defaults` in Ax (with a class method it's very contained and clear what needs to be done: convert inputs A to input B, whereas with a utility there is going to be an ever-growing tree of decisions and shared logic that easily grows unparse-able, case in point: https://github.com/facebook/Ax/blob/5547f8b8fdb7890731e10062b27a24e7abebc395/ax/models/torch/botorch_defaults.py#L128).

> This is kind of hard to read on the Ax side too because many things are also passed around via **kwargs

This might also creep up in the utility case, I think? We might want to try to avoid that as much as we can whether we go for a utility or class method, this point makes a lot of sense.

A utility also has an added benefit (hopefully) that it's a bit harder to accidentally enable `construct_inputs` for a given class that way; a class method could be inherited, but adding a specific class to a utility presumably requires a conscious decision?

**Reply by Balandat** [reply]
OK how about a registry that we auto-populate by means of a decorator? We can key directly on the class, that way there shouldn't really be any big if/else trees. Sth like the following:

```
from typing import Type, Callable, Any

from botorch.acquisition.acquisition import AcquisitionFunction
from botorch.acquisition.analytic import ExpectedImprovement
from botorch.acquisition.monte_carlo import qUpperConfidenceBound


ACQF_INPUT_CONSTRUCTOR_REGISTRY = {}


def acqf_input_constructor(
    acqf_class: Type[AcquisitionFunction]
) -> Callable[[], AcquisitionFunction]:
    if acqf_class in ACQF_INPUT_CONSTRUCTOR_REGISTRY:
        raise ValueError(f"Cannot register duplicate arg constructor for acquisition class ({acqf_class})")

    def decorator(method):
        ACQF_INPUT_CONSTRUCTOR_REGISTRY[acqf_class] = method
        return method

    return decorator


@acqf_input_constructor(ExpectedImprovement)
def construct_inputs_expected_improvement(*args, **kwargs):
    # TODO: Real stuff
    return {"model": "model", "best_f": args[0]}


@acqf_input_constructor(qUpperConfidenceBound)
def construct_inputs_q_upper_confidence_bound(*args, **kwargs):
    # TODO: Real stuff 
    return {"model": "model", "beta": args[0]}

```

If we also wrote sth simple like

```
def get_acqf_input_constructor(acqf_class: Type[AcquisitionFunction]):
    if acqf_class not in ACQF_INPUT_CONSTRUCTOR_REGISTRY:
        raise RuntimeError(
            "Input constructor for acquisition class {acqf_class} not registered. "
            "Use the `@acqf_input_constructor` decorator to register a new method."
        ) 
    return ACQF_INPUT_CONSTRUCTOR_REGISTRY[acqf_class]
```

then upstream we could use this as follows
```
from botorch.acquisition.utils import get_acqf_input_constructor

,,,

input_arg_constructor = get_acqf_input_constructor(botorch_acqf_class)
acqf = acqf_class(**input_arg_constructor(training_data, model, ...))
```


Thoughts?

**Reply by lena-kashtelyan** [reply]
Love this solution, it's very creative and seems like the best of all the worlds.

**Reply by sdaulton** [reply]
Yeah I like this solution. I am definitely in favor of keeping the construct_inputs logic outside of the core acquisition classes. This is an elegant solution

**Reply by Balandat** [reply]
sg, will put up a PR for this

**Reply by Balandat** [reply]
PR is here: #788
